File: Software/Golden Measure/LSB_decoder.py
"""
	
    This program is a Golden Measure for the Image Steganography Detector (ISD). 
    The program does not process every pixel but processes 8 LSBs at a time until the delimeter is found. It then returns the message and writes it to a text file.
    The remaining pixels are not processed.

"""

# imports - reading in hex file
from intelhex import IntelHex
import time

# LSB Decoder
def LSB_decoder():
    # variables
    lsbs_combined = ''
    character = ''
    out = ''
    count = 0

    # start performance timing
    start = time.perf_counter()

    # open file with hex values
    f = open("stego_2500.hex", "r")

    # loop through each line of hex file
    for line in f:
        # convert hex value into binary
        b = bin(int(line, 16)).zfill(8)

        # retrieve and save LSBs
        LSB = str((b[len(b) - 1]))
        lsbs_combined = lsbs_combined + LSB

        # counter tp keep track of number of bits collected
        count += 1

        # once a byte is collected, convert to char
        if (count == 8):
            # convert byte into character
            val = int(lsbs_combined, 2)
            character = chr(val)

            # if character is the delimeter break
            if (character == '$'):
                break
            
            # save character string
            out = out + str(character)

            # reset count and lsbs_combined
            count = 0
            lsbs_combined = ''

    # end performance timing
    end = time.perf_counter()
    
    # output message
    print(out+"\n")

    # save output to txt
    with open('PY_message.txt', 'w') as f:
        f.write(out)
        print("Saved to PY_message.txt \n")

    # print timing
    # print(f"Decoded the message and saved to text file in {end - start:0.4f} seconds\n")
    print(f"{end - start}")

def main():
    LSB_decoder()

if __name__ == "__main__":
    main()

File: Software/Golden Measure/LSB_decoder.c
#include <stdio.h>
#include <string.h>
 #include <time.h>

int numBytes = 150000;

int main(){
    /********* TIMING ************/
    clock_t start, end;
    double cpu_time_used;
    
    start = clock(); // START TIMING

    /*****************************/

    int numBits = 8*numBytes;
    /* READ IN THE CONTENTS OF THE .hex FILE*/
    FILE *hfile;

    // Open file in read mode
    hfile = fopen("stego_2500.hex", "r");

    // If the file does not exist, print error message and terminate program
    if(hfile == NULL) {
        printf("Not able to open the file.");
        return 1;
    }

    /* STORING FILE CONTENTS IN ARRAY */
    char myString[1000];
    char message[numBits];
    int lsb[8];
    int count = 0;

    // Read the contents of the file and retrieve LSBs
    while(fgets(myString, 1000, hfile)) {
        //printf("%s",myString);

        sscanf(myString,"%x",(lsb+count));
        *(lsb+count) = *(lsb+count)&1; //get LSB and store in the array

        //printf("%d\n",*(lsb+count));

        /* GROUP LSBS INTO BYTES TO DECODE MESSAGE */ 
        if(count==8){
            int decimal = 0;
            char character;

            // bit shift LSBs stored in array to store byte as binary integer
            decimal = lsb[0]<<7 | lsb[1]<<6 | lsb[2] <<5 | lsb[3] <<4 | lsb[4] <<3 | lsb[5] <<2 | lsb[6] <<1 | lsb[7] ;
            //printf("\nThe decimal value is: %d\n",decimal);

            // convert binary integer to character
            character = (char) decimal;
            //printf("The char is: %c\n",character);

            // if the delimeter is found, stop processing
            if(character=='$'){ 
                break;
            }

            // add characters to message array
            sprintf(message, "%s%c", message, character);  

            // reset count to 0
            count = 0;
        }

        // increment count for each LSB retrieved
        count=count+1;
    }

    // Close the file
    fclose(hfile);

    /* WRITE RECOVERED MESSAGE TO FILE*/

    FILE *hfile2;

    // Try to open file
    hfile2 = fopen("c_message.txt","w");

    // If the file cannot be opened, print error message and terminate program
    if(hfile2 == NULL) {
        printf("Not able to open the output file.");
        return 1;
    }

    // Write message to file
    fputs(message,hfile2);

    // Close file
    fclose(hfile2);

    /********* TIMING ************/
    end = clock(); // END TIMING

    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC; // Convert ticks to seconds
    //printf("Time %f seconds\n",cpu_time_used);

    printf("%f\n", cpu_time_used);
    /****************************/
    
}